standard/basic-concepts.md

```diff
@@ -78,12 +78,16 @@ There are several different types of declaration spaces, as described in the fol
-- Each *block* or *switch_block*, as well as a `for`, `foreach`, and `using` statement, creates a local variable declaration space for local variables and local constants. Names are introduced into this declaration space through *local_variable_declaration*s and *local_constant_declaration*s. Note that blocks that occur as or within the body of a function member, anonymous function, or local function are nested within the local variable declaration space declared by those functions for their parameters. Thus, it is an error to have, for example, a method with a local variable and a parameter of the same name.
-
+- Each method declaration, property declaration, property accessor declaration, indexer declaration, indexer accessor declaration, operator declaration, instance constructor declaration, anonymous function, and local function creates a new declaration space called a ***local variable declaration space***. Names are introduced into this declaration space through formal parameters (*fixed_parameter*s and *parameter_array*s) and *type_parameter*s. The set accessor for a property or an indexer introduces the name `value` as a formal parameter.
+- Additional local variable declaration spaces may occur within member declarations, anonymous functions and local functions. Names are introduced into these declaration spaces through *declaration_expression*s and *declaration_statement*s. Local variable declaration spaces may be nested, but it is an error for a local variable declaration space and a nested local variable declaration space to contain elements with the same name. Thus, within a nested declaration space it is not possible to declare a local variable or constant with the same name as a parameter, type parameter, local variable or constant in an enclosing declaration space. It is possible for two declaration spaces to contain elements with the same name as long as neither declaration space contains the other. Local declaration spaces are created by the following constructs:
+  - Each *variable_initializer* in a field and property declaration introduces its own local variable declaration space, that is not nested within any other local variable declaration space.
+  - The body of a function member, anonymous function, or local function, if any, creates a local variable declaration space that is considered to be nested within the function's local variable declaration space.
+  - Each *constructor_initializer* creates a local variable declaration space nested within the instance constructor declaration. The local variable declaration space for the constructor body is in turn nested within this local variable declaration space.
+  - Each *block*, *switch_block*, *iteration_statement* and *using_statement* creates a nested local variable declaration space.
+  - Each *embedded_statement* that is not directly part of a *statement_list* creates a nested local variable declaration space.
```

Comment:
*Pattern variables*: I left those out for now. Let's discuss which PR should do them.
*Switch guards*: Same.
*Query clauses*: We've generally taken the stance that query clauses are defined by their desugaring. 
*Catch blocks*: Good catch! It's orthogonal to the subject of this PR. I can add it in anyway, or we can open an issue. What do you think?

Reply:
Query clauses define most of their their semantics by the translation, but the scoping is not based on that translation.  These scopes only came into existence in C# 7 because of its expression variables.

I think it is something like this:
- Variables introduced in the expression of a `from` clause are in scope for the whole query expression introduced by that `from` clause.
- For other clauses, each expression introduces its own scope that contains only that expression.

Reply:
Captured in #768 to be addressed post-merge.